### ISSUE TITLE
Failed refreshing ICE server configuration with status code 0x0000000f

### ISSUE THREAD

**Issue by david511** [awslabs/amazon-kinesis-video-streams-webrtc-sdk-c#1270]
After about 10 hours running kvs webrtc v1.6.0 on an EC2 instance, viewers are not able to connect to the signaling channel. Attached are the logs which indicates why the disconnection happens. Could this have anything to do with the AWS session token expiring? I am able to connect to the signaling channel fine within the 10 hour period, anything after that I can't.

2021-09-13 23:42:07 VERBOSE lwsHttpCallbackRoutine(): HTTPS callback with reason 1
2021-09-13 23:42:07 WARN    lwsHttpCallbackRoutine(): Client connection failed. Connection error string: closed before established
2021-09-13 23:42:07 VERBOSE lwsHttpCallbackRoutine(): HTTPS callback with reason 30
2021-09-13 23:42:07 VERBOSE lwsWssCallbackRoutine(): WSS callback with reason 10
2021-09-13 23:42:07 DEBUG   lwsWssCallbackRoutine(): Client is writable
2021-09-13 23:42:07 ERROR   refreshIceConfiguration(): operation returned status code: 0x0000000f
2021-09-13 23:42:07 WARN    signalingClientError(): Signaling client generated an error 0x5d00002f - 'Failed refreshing ICE server configuration with status code 0x0000000f.'
2021-09-13 23:42:07 ERROR   signalingGetIceConfigInfoCout(): operation returned status code: 0x0000000f
2021-09-13 23:42:07 ERROR   initializePeerConnection(): operation returned status code: 0x0000000f
2021-09-13 23:42:07 DEBUG   freeSampleStreamingSession(): Freeing streaming session with peer id: 78e7bc0a-cd64-4287-bdbe-8ec632f8b6f4
2021-09-13 23:42:07 ERROR   closePeerConnection(): operation returned status code: 0x00000001
2021-09-13 23:42:07 ERROR   freeSampleStreamingSession(): operation returned status code: 0x00000001
2021-09-13 23:42:07 ERROR   signalingMessageReceived(): operation returned status code: 0x0000000f
2021-09-13 23:42:07 ERROR   receiveLwsMessageWrapper(): operation returned status code: 0x0000000f
2021-09-13 23:42:07 INFO    freeSignalingClient(): Freeing Signaling Client
2021-09-13 23:42:07 VERBOSE lwsWssCallbackRoutine(): WSS callback with reason 10
2021-09-13 23:42:07 VERBOSE lwsWssCallbackRoutine(): WSS callback with reason 75
2021-09-13 23:42:07 VERBOSE lwsHttpCallbackRoutine(): HTTPS callback with reason 30
2021-09-13 23:42:07 VERBOSE lwsHttpCallbackRoutine(): HTTPS callback with reason 28
2021-09-13 23:42:07 VERBOSE lwsWssCallbackRoutine(): WSS callback with reason 28
2021-09-13 23:42:07 INFO    createSignalingClientSync(): Creating Signaling Client Sync
2021-09-13 23:42:07 ERROR   signalingCacheLoadFromFile(): operation returned status code: 0x00000002
2021-09-13 23:42:07 WARN    createSignalingSync(): Failed to load signaling cache from file
2021-09-13 23:42:07 VERBOSE lwsHttpCallbackRoutine(): HTTPS callback with reason 21
2021-09-13 23:42:07 VERBOSE signalingClientStateChanged(): Signaling client state changed to 1 - 'New'
2021-09-13 23:42:07 DEBUG   stepStateMachine(): State Machine - Current state: 0x0000000000000001, Next state: 0x0000000000000002
2021-09-13 23:42:07 VERBOSE signalingClientStateChanged(): Signaling client state changed to 2 - 'Get Security Credentials'
2021-09-13 23:42:07 DEBUG   stepStateMachine(): State Machine - Current state: 0x0000000000000002, Next state: 0x0000000000000004
2021-09-13 23:42:07 VERBOSE signalingClientStateChanged(): Signaling client state changed to 3 - 'Describe Channel'
2021-09-13 23:42:07 VERBO[2021/09/13 23:42:07:8574] E: getaddrinfo failed: -2
[2021/09/13 23:42:08:0576] E: getaddrinfo failed: -2
[2021/09/13 23:42:08:4579] E: getaddrinfo failed: -2
SE lwsCompleteSync(): Perform secure synchronous call for URL: https://kinesisvideo..amazonaws.com/describeSignalingChannel

**Comment by MushMal** [comment]
Seems you have no network connection. Check for the network.

**Comment by david511** [comment]
Network connection is fine. If I restart the executable I am able to reconnect to the signaling channel but again after ~10 hours I get the same error. I looked into the error codes and 0x5d000002f means "STATUS_SIGNALING_ICE_CONFIG_REFRESH_FAILED", 0x000000f means operation timeout and the error code "E: getaddrinfo failed: -2" means name or service are unknown. Could this have something to do with the libwebsocket library? Are there any other variables within kvs I could modify such as timeout intervals to help narrow down this issue?

**Comment by MushMal** [comment]
Might be some sort of a fd or handle leak perhaps. Are you using the latest checked in code? I believe the version of libWebSockets had been updated recently. 

What's strange is that both the libWebSocket based operations are failing as well as the raw socket operations. Something at the OS socket level is not working properly.

**Comment by david511** [comment]
I am using the code from this commit 8b8b2bdf064f6cb2b6495339d31efc3518b12eb9. Has it been updated after that?

**Comment by MushMal** [comment]
yes, there were multiple changes from Feb 15th. Try to debug your network connection when you get to this scenario - is it possible that there are firewall rules or anything else comes into picture. Try to collect more data and use various setups to try to identify the root cause. 

Change-wise, the only one that comes to my mind is the latest bump to the libWebSockets version

**Comment by david511** [comment]
So would upgrading libWebSockets version also mean i would need to upgrade the kvs library? If so, Is there a stable version of both not too far off from feb 15th?

**Comment by david511** [comment]
Checked the network connection, all is fine with that and there are no firewall rules or anything like that. From my understanding the libWebSockets got upgraded at this commit 374f025cee2981462c04a7b7a584775ce8a5578c. Would you recommend just upgrading the kvs webrtc library to version 1.6.0 or do you think it is okay to just checkout out a commit right after you guys upgraded the libWebSockets library?

There has been a lot of testing with my app so I am trying to avoid a complete rewrite of the integration with my app and the kvs webrtc library.

**Comment by MushMal** [comment]
The changes should be backward compatible. You can test out the changes on the stable release commits.

**Comment by david511** [comment]
I went ahead and upgraded the library however same issue occurred but I got a more detailed error message.

2021-09-21 13:56:51 WARN    lwsHttpCallbackRoutine(): Client connection failed. Connection error string: DNS NXDOMAIN
2021-09-21 13:56:51 VERBOSE lwsHttpCallbackRoutine(): HTTPS callback with reason 30
2021-09-21 13:56:51 VERBOSE signalingClientStateChanged(): Signaling client state changed to 3 - 'Describe Channel'
2021-09-21 13:56:54 VERBOSE lwsCompleteSync(): Perform secure synchronous call for URL: https://kinesisvideo..amazonaws[2021/09/21 13:56:54:8085] N:  ++ [wsicli|5|POST/h1/kinesisvideo..amazonaws.com] (1)
[2021/09/21 13:56:54:8085] N:  -- [wsicli|5|POST/h1/kinesi

The issue appears to be that the DNS failed "DNS NXDOMAIN". After doing some research AWS EC2 DNS server can fail if there are too many DNS queries. Does the kvs webrtc/libwebsockets send a lot of queries? Do you think this could be the issue?

**Comment by MushMal** [comment]
KVS WebRTC client has to make a few calls - all well within the TPS limits. For example, it needs to make a call to get the security token (exchanging the cert for a token for example), it needs to make a describe call, get data endpoint, get ICE configuration and create/open the signaling channel. With each call failing, it can be retried in an exponential callbacks up-to 5 times with increasing wait time in between the calls. I am not sure if the DNS TPS is your issue. Is there a way to debug your connection on the EC2 instance? Perhaps contact Amazon about it?

**Comment by david511** [comment]
I will look into debugging the network connection tomorrow. It looks like kvs webrtc stops at state 3 "2021-09-21 19:47:22 VERBOSE signalingClientStateChanged(): Signaling client state changed to 3 - 'Describe Channel'
2021-09-21 19:47:22 VERBOSE lwsCompleteSync(): Perform secure synchronous call for URL: https://kinesisvideo..amazonaws.com/describeSignalingChannel"

Right after that is when the DNS NXDOMAIN error appears. What is that URL? is it truncated for logging purposes? Do you have an example of what this URL should look like?

**Comment by MushMal** [comment]
The URI will be combined taking into account your region.

The docs for the API are here: https://docs.aws.amazon.com/kinesisvideostreams/latest/dg/API_DescribeSignalingChannel.html

**Comment by david511** [comment]
What would cause the library not being able to get the region or ARN before making a call to the describe signaling channel url? Network connection and DNS appear to be fine after debugging it. I even tested it with a 8.8.8.8 DNS nameserver.

**Comment by MushMal** [comment]
Try to build the debug version of libWebSockets to see what's in the debug spew

**Comment by david511** [comment]
i am running the debug version now, where can I find the debug spew? Should it be in stdout?

**Comment by MushMal** [comment]
Generally, I haven't seen DNS NXDOMAIN errors before. This is indicative of your DNS not resolving. Is there anything special in your setup? Can you run this locally? Can you run this on a different vm image? Which region are you using? getaddrinfo is a client-side error that can be caused by the DNS resolution failures. I don't think you are going to get much help on this forum so I would recommend getting more info on your scenario.

**Comment by david511** [comment]
i tested it on multiple instances and locally but still get the same error. When i ran on my local computer the error actually happened within 2 hours. There is nothing special about the set up. The region we are using is us-east-1. I can see it is failing to describe the signaling channel in signaling client state 3. This is the URL i was asking about earlier. Would it be worth my effort to see if I can somehow cache the describe channel URL with the ARN and region when it is working and then use that cached URL from then onwards if that makes sense?

**Comment by MushMal** [comment]
The default configuration in the samples does the caching itself - you could change the TTL of the cache to your needs. I am just not sure what could cause this - especially, since this is as you said reproducible on multiple devices - assume also multiple networks as the network itself in question could cause this issue. Perhaps someone from AWS could have a better idea?

**Comment by david511** [comment]
I actually am using a different cert.pem file then the default one that comes with the repo. Would that make a difference?

**Comment by MushMal** [comment]
I am actually really surprised it worked without specifying the cert with the fingerprint highlighted here: https://www.amazontrust.com/repository/SFSRootCAG2.pem

In general, any issues you might have in this complex environment, you debug by a deductive method by getting the sample working on your environment and only then modifying bits at a time.

**Comment by david511** [comment]
@MushMal After doing more debugging I can see the libwebsockets library is returning a message saying "The security token included in the request is expired".
[2021/09/27 16:51:58:1600] D: 0000: 7B 20 22 6D 65 73 73 61 67 65 22 3A 20 22 54 68    { "message": "Th
[2021/09/27 16:51:58:1600] D: 0010: 65 20 73 65 63 75 72 69 74 79 20 74 6F 6B 65 6E    e security token
[2021/09/27 16:51:58:1600] D: 0020: 20 69 6E 63 6C 75 64 65 64 20 69 6E 20 74 68 65     included in the
[2021/09/27 16:51:58:1600] D: 0030: 20 72 65 71 75 65 73 74 20 69 73 20 65 78 70 69     request is expi
[2021/09/27 16:51:58:1600] D: 0040: 72 65 64 22 20 7D                                  red" }
[2021/09/27 16:51:58:1600] D

This message appears right after lwsHttpCallbackRoutine() appends x-amz-security-token to an http request.
Have you ever come across this message? If so is there an easy was to resolve this? Do you think it is likely this is the root issue i am facing?

**Comment by MushMal** [comment]
I have seen this with some direct fault injection. Perhaps you could explain how you integrate the auth? There is a possibility that the token doesn't get refreshed due to time skew/drift. 

* Make sure you are not changing the caching algorithm of the auth provider
* Make sure you have a way to synchronize your device time - perhaps using NTP client of some sort at boot. 

Auth module is pretty much the only thing that requires a correct absolute time.

**Comment by david511** [comment]
I have attached an IAM role to my ec2 instance and I perform a call to the instances metadata using this URL "http://169.254.169.254/latest/meta-data/iam/security-credentials/<iam_role_name>" to get the AWS access key, secret key and session token. Then, i do the same thing to get the region but with a difference URL. After this is done I give the credentials to createSampleConfiguration() which calls createStaticCredentialProvider(). This IAM role session duration is set to an hour however it takes about 7-10 hours before the kvs webrtc gets this error/log message and stops working. The output of timedatectl status is:
"Local time: Mon 2021-09-27 19:49:59 UTC
Universal time: Mon 2021-09-27 19:49:59 UTC
RTC time: Mon 2021-09-27 19:49:59
Time zone: n/a (UTC, +0000)
System clock synchronized: no
systemd-timesyncd.service active: no
RTC in local TZ: no"

When I run on my local machine I am using "Aws::STS::STSClient" to get the session token from an Assume role. Again once i have all the credential information I just pass it to createSampleConfiguration(). I have found the when I am running on my local machine it fails within 2-3 hours vs when I run on an EC2 instance. My local machine is NTP synchronized. I have verify this by running "timedatectl status". Output below:
"Local time: Mon 2021-09-27 19:50:13 UTC
Universal time: Mon 2021-09-27 19:50:13 UTC
Timezone: Etc/UTC (UTC, +0000)
NTP enabled: yes
NTP synchronized: yes
RTC in local TZ: no
DST active: n/a"

I have not edited any caching auth provider related stuff. I am assuming that kvs webrtc should be refreshing credentials if they need to be refreshed, is this correct? The region I am using is us-east-1 but the timezome of the instance/local computer is UTC according to the "date" command.

**Comment by MushMal** [comment]
This is likely an internal behavior and one that you should not design against. You SHOULD NOT be using static credential provider in any devices anyway. In your case, you should have a rotating credential - similar to an IoT credential provider or file credential provider which would update the credentials when they go into "grace" period. You can design your credential provider following exact steps as the IoT credential provider which is in Common library.

**Comment by david511** [comment]
Okay I will try to implement the IoT credential provider. Just to clarify, by using the IoT credential provider it should automatically refresh the security token before it is going to expire and with the static credential provider it never gets refreshed, Correct?

**Comment by MushMal** [comment]
Just to be precise... You could use IoT credential provider as is if your application calls for IoT integration. However, if you are using the SDK on the VM for your scenario, then perhaps you could copy and modify the IoT credential provide (use it as a template) but replace the calls to the IoT backend and the response parsing to your own.

The IoT and File-based credential providers do indeed handle the token rotation without any need for the application to intervene.

**Comment by david511** [comment]
Okay I am doing tests with the IoT credential provider now, seems to be working good so far. I also want to test out file credential provider but keep getting error code 0x15000024 which is a bad file format error. Should the file format be the same as in ~/.aws/credentials?

**Comment by david511** [comment]
Okay I am doing tests with the IoT credential provider now, seems to be working good so far. I also want to test out file credential provider but keep getting error code 0x15000024 which is a bad file format error. Should the file format be the same as in ~/.aws/credentials? If not do you have an example of the file format or know where I can find one?

**Comment by MushMal** [comment]
There really is a lack of tests as I couldn't find one for FileAuthCallbacks in C Producer or FileCredentialProvider  in both C producer and WebRTC. @disa6302 perhaps this is something you guys might want to look at.

@david511 you can take a look at this line of code:

https://github.com/awslabs/amazon-kinesis-video-streams-producer-c/blob/master/src/source/Common/FileCredentialProvider.c#L157

This is the code that reads a line from the file - space separated. Really not very sophisticated. The usage is that there is some other entity that keeps updating the file whereas the SDK would read it up before the credentials expire for the new set.

**Comment by MushMal** [comment]
The error states that your reconnecting failed with the underlying error of STATUS_SIGNALING_GET_TOKEN_CALL_FAILED. Your logs don't indicate anything about why this failed - perhaps those are incomplete.

File-based credential provider (as you can see in the code) will only read from the specified file just before the expiration of the existing credentials. It's up to your code to ensure you keep updating this file periodically - slightly earlier than the file will be checked. Ex: If your session token expires in 1 hour, try to refresh and update the file in 45 mins.

**Comment by david511** [comment]
I update the file every 45 minutes however I still get the same error. I only get this error when I try to connect to the signaling channel after 45-60 minutes. If I connect anytime before that it works fine. My credential file contains the access, secret and session token as well as an expiration value of 100ns. Does it have to do anything with the type of credentials I am using? I am using an IAM role and retrieving the credentials via the ec2 metedata. From my understanding the token rotation rate is every 40 minutes however I also changed that to every 2 minutes and did a test which didn't change anything. When I reconnect to the signaling channel this is the log I get:

01 00:43:08 VERBOSE lwsWssCallbackRoutine(): WSS callback with reason 9
2021-10-01 00:43:08 VERBOSE signalingClientGetCurrentState(): Signaling Client Get Current State
2021-10-01 00:43:08 VERBOSE lwsWssCallbackRoutine(): WSS callback with reason 8
2021-10-01 00:43:08 VERBOSE lwsWssCallbackRoutine(): WSS callback with reason 8
2021-10-01 00:43:08 DEBUG   receiveLwsMessage(): Client received message of type: SDP_OFFER
2021-10-01 00:43:08 VERBOSE lwsWssCallbackRoutine(): WSS callback with reason 10
2021-10-01 00:43:08 DEBUG   lwsWssCallbackRoutine(): Client is writable
2021-10-01 00:43:08 INFO    signalingClientGetIceConfigInfoCount(): Signaling Client Get ICE Config Info Count
2021-10-01 00:43:08 DEBUG   refreshIceConfiguration(): Refreshing the ICE Server Configuration
2021-10-01 00:43:08 DEBUG   stepStateMachine(): State Machine - Current state: 0x0000000000000100, Next state: 0x0000000000000020
2021-10-01 00:43:08 VERBOSE signalingClientStateChanged(): Signaling client state changed to 6 - 'Get ICE Server Configuration'
2021-10-01 00:43:08 DEBUG   stepStateMachine(): State Machine - Current state: 0x0000000000000020, Next state: 0x0000000000000002
2021-10-01 00:43:08 VERBOSE signalingClientStateChanged(): Signaling client state changed to 2 - 'Get Security Credentials'
2021-10-01 00:43:08 VERBOSE readFileCredentials(): Reading AWS credentials from file: /storage/deastwood/ort/creds.txt, file length = 1137.
2021-10-01 00:43:08 VERBOSE signalingClientStateChanged(): Signaling client state changed to 2 - 'Get Security Credentials'
2021-10-01 00:43:08 VERBOSE readFileCredentials(): Reading AWS credentials from file: /storage/deastwood/ort/creds.txt, file length = 1137.
2021-10-01 00:43:08 VERBOSE signalingClientStateChanged(): Signaling client state changed to 2 - 'Get Security Credentials'
2021-10-01 00:43:08 VERBOSE readFileCredentials(): Reading AWS credentials from file: /storage/deastwood/ort/creds.txt, file length = 1137.
2021-10-01 00:43:08 VERBOSE signalingClientStateChanged(): Signaling client state changed to 2 - 'Get Security Credentials'
2021-10-01 00:43:08 VERBOSE readFileCredentials(): Reading AWS credentials from file: /storage/deastwood/ort/creds.txt, file length = 1137.
2021-10-01 00:43:08 VERBOSE signalingClientStateChanged(): Signaling client state changed to 2 - 'Get Security Credentials'
2021-10-01 00:43:08 VERBOSE readFileCredentials(): Reading AWS credentials from file: /storage/deastwood/ort/creds.txt, file length = 1137.
2021-10-01 00:43:08 VERBOSE signalingClientStateChanged(): Signaling client state changed to 2 - 'Get Security Credentials'
2021-10-01 00:43:08 VERBOSE readFileCredentials(): Reading AWS credentials from file: /storage/deastwood/ort/creds.txt, file length = 1137.
2021-10-01 00:43:08 VERBOSE signalingClientStateChanged(): Signaling client state changed to 2 - 'Get Security Credentials'
2021-10-01 00:43:08 VERBOSE readFileCredentials(): Reading AWS credentials from file: /storage/deastwood/ort/creds.txt, file length = 1137.
2021-10-01 00:43:08 VERBOSE signalingClientStateChanged(): Signaling client state changed to 2 - 'Get Security Credentials'
2021-10-01 00:43:08 VERBOSE readFileCredentials(): Reading AWS credentials from file: /storage/deastwood/ort/creds.txt, file length = 1137.
2021-10-01 00:43:08 VERBOSE signalingClientStateChanged(): Signaling client state changed to 2 - 'Get Security Credentials'
2021-10-01 00:43:08 VERBOSE readFileCredentials(): Reading AWS credentials from file: /storage/deastwood/ort/creds.txt, file length = 1137.
2021-10-01 00:43:08 VERBOSE signalingClientStateChanged(): Signaling client state changed to 2 - 'Get Security Credentials'
2021-10-01 00:43:08 VERBOSE readFileCredentials(): Reading AWS credentials from file: /storage/deastwood/ort/creds.txt, file length = 1137.
2021-10-01 00:43:08 VERBOSE signalingClientStateChanged(): Signaling client state changed to 2 - 'Get Security Credentials'
2021-10-01 00:43:08 VERBOSE readFileCredentials(): Reading AWS credentials from file: /storage/deastwood/ort/creds.txt, file length = 1137.
2021-10-01 00:43:08 ERROR   refreshIceConfiguration(): operation returned status code: 0x5d000002
2021-10-01 00:43:08 WARN    signalingClientError(): Signaling client generated an error 0x5d00002f - 'Failed refreshing ICE server configuration with status code 0x5d000002.'
2021-10-01 00:43:08 ERROR   signalingGetIceConfigInfoCout(): operation returned status code: 0x5d000002
2021-10-01 00:43:08 ERROR   initializePeerConnection(): operation returned status code: 0x5d000002
2021-10-01 00:43:08 DEBUG   freeSampleStreamingSession(): Freeing streaming session with peer id: 798ipjg5r46d2uq2jaqhyq 
2021-10-01 00:43:08 ERROR   closePeerConnection(): operation returned status code: 0x00000001
2021-10-01 00:43:08 ERROR   freeSampleStreamingSession(): operation returned status code: 0x00000001
2021-10-01 00:43:08 ERROR   signalingMessageReceived(): operation returned status code: 0x5d000002
2021-10-01 00:43:08 ERROR   receiveLwsMessageWrapper(): operation returned status code: 0x5d000002
2021-10-01 00:43:08 INFO    freeSignalingClient(): Freeing Signaling Client

**Comment by MushMal** [comment]
As before, the log snippet you've attached doesn't indicate why PIC or Signaling code rejects your credentials (assuming they have been updated). Are you sure you are updating the absolute value of the expiration? Can you run some experiments and add debug printing for you to understand the expiration and why this fails?

**Comment by david511** [comment]
what do you meaning updating the expiration, I thought it should just be a fixed value? Am I supposed to incrementing it every time I update the file?

**Comment by MushMal** [comment]
It denotes not a period but rather an absolute time. NOTE: as mentioned before, this is the only time that should be sync-ed with the backend service.

**Comment by david511** [comment]
so the expiration time needs to always be set greater than the current time. When I say current time I am referring to the value returned from GETTIME() in the kvs library. The expiration value in the file should be expiration = currentTime + 30 minutes for example. Is this accurate?

**Comment by MushMal** [comment]
It should be greater than the current time + min period. I believe min period is set to something like a minute or so. Note that the GETTIME() implementation by default is calling into the platforms clock. This clock should be in-sync (with some delta) with the backend server (which sync with NTP).

**Comment by david511** [comment]
Do you know what the macro is called for min period? What exactly is the min period referring to, the minimum period the credentials need to be valid for? I am assuming this is the reason why I am facing the above error as I must have been giving it the wrong absolute time value.

**Comment by MushMal** [comment]
This is where the check is done in file credential provider:

https://github.com/awslabs/amazon-kinesis-video-streams-producer-c/blob/master/src/source/Common/FileCredentialProvider.c#L128

The definition of the grace is here: https://github.com/awslabs/amazon-kinesis-video-streams-producer-c/blob/master/src/source/Common/FileCredentialProvider.h#L14

If you chase the values, it comes down to 45 seconds. 

To sum it up... Your new credentials expiration should be an absolute time in the future which is over 45 seconds into the future.

**Comment by disa6302** [comment]
Closing because we have an open PR.

**Comment by david511** [comment]
@disa6302 just out of curiosity, do you have the link to that PR?
